compiler/src/dotty/tools/dotc/transform/TypeTestsCasts.scala

```diff
@@ -244,6 +244,13 @@ object TypeTestsCasts {
           else if (isDerivedValueClass(testCls)) {
             expr // adaptToType in Erasure will do the necessary type adaptation
           }
+          else if (testCls eq defn.NothingClass) {
+            // In the JVM `x.asInstanceOf[Nothing]` would throw a class cast exception except when `x eq null`.
+            // To avoid this loophole we execute `x` and then regardless of the result throw a `ClassCastException`
```

Comment:
For completeness I would add a test where x is a side effect

Reply:
Added